c2rust-analyze/src/rewrite/expr/hir_op.rs

```diff
@@ -244,47 +263,112 @@ impl<'a, 'tcx> intravisit::Visitor<'tcx> for HirRewriteVisitor<'a, 'tcx> {
         // the rewrite should occur at the callsite
         let callsite_span = ex.span.source_callsite();
 
-        if let Some(loc) = self.find_primary_location(ex) {
-            self.locations_visited.insert(loc);
-            let rws = self.rewrites.get(&loc).map_or(&[] as &[_], |v| v);
-            for rw in rws {
-                hir_rw = match rw.kind {
-                    mir_op::RewriteKind::OffsetSlice { mutbl } => {
-                        // `p.offset(i)` -> `&p[i as usize ..]`
-                        assert!(matches!(hir_rw, Rewrite::Identity));
-                        let arr = self.get_subexpr(ex, 0);
-                        let idx =
-                            Rewrite::Cast(Box::new(self.get_subexpr(ex, 1)), "usize".to_owned());
-                        let elem = Rewrite::SliceTail(Box::new(arr), Box::new(idx));
-                        Rewrite::Ref(Box::new(elem), mutbl_from_bool(mutbl))
-                    }
-
-                    mir_op::RewriteKind::SliceFirst { mutbl } => {
-                        // `p` -> `&p[0]`
-                        let arr = hir_rw;
-                        let idx = Rewrite::LitZero;
-                        let elem = Rewrite::Index(Box::new(arr), Box::new(idx));
-                        Rewrite::Ref(Box::new(elem), mutbl_from_bool(mutbl))
-                    }
-
-                    mir_op::RewriteKind::MutToImm => {
-                        // `p` -> `&*p`
-                        let place = Rewrite::Deref(Box::new(hir_rw));
-                        Rewrite::Ref(Box::new(place), hir::Mutability::Not)
-                    }
-
-                    mir_op::RewriteKind::RemoveAsPtr => {
-                        // `slice.as_ptr()` -> `slice`
-                        assert!(matches!(hir_rw, Rewrite::Identity));
-                        self.get_subexpr(ex, 0)
-                    }
-
-                    mir_op::RewriteKind::RawToRef { mutbl } => {
-                        // &raw _ to &_ or &raw mut _ to &mut _
-                        Rewrite::Ref(Box::new(self.get_subexpr(ex, 0)), mutbl_from_bool(mutbl))
-                    }
-                };
+        let locs = self.find_locations(ex);
+        for loc in &locs {
+            self.locations_visited.insert(*loc);
+        }
+
+        let all_rws_unflattened: Vec<_> = locs
+            .iter()
+            .filter_map(|loc| self.rewrites.get(loc))
+            .map(|rws| rws.iter().map(|rw| &rw.kind).collect::<Vec<_>>())
+            .collect();
+        let all_rws_unflattened = all_rws_unflattened
+            .iter()
+            .map(|v| v.as_slice())
+            .collect::<Vec<_>>();
```

Comment:
This might be clearer without the filter, so that a list of N locations turns into N lists of rewrites (one from each location), where some of those lists might be empty.

Reply:
Without the filter there are cases such as `[[], [RawToRef { mutbl: false }]]` to match against. I'm not sure we want to match against specific cases here, and we can instead do

```rust
             rws => rws
                    .iter()
                    .flat_map(|rws| rws.iter())
                    .fold(hir_rw, |acc, rw| rewrite_from_mir_rws(rw, acc))
```
but this matches against all cases and presumably we want to panic in some cases, no?

Reply:
>Without the filter there are cases such as `[[], [RawToRef { mutbl: false }]]` to match against.

This is just the sort of thing I was thinking of in https://github.com/immunant/c2rust/pull/893#discussion_r1177062926 above: we expect to see two `Location`s, one with a `Ref` and the other with an `AddressOf` (checked by `is_addr_of_expansion`), where the first has no rewrites and the second has only `RawToRef` (checked by this pattern).

Reply:
Good point, I missed that. I've made the change in 8114679a, but there is one other type of instance where the compiler will generate these two ref/raw MIR statements:

```rust
let r: *mut request_st = &mut (*con).request;
```

which doesn't occur in a macro call. I've accounted for this case in the match arm